### ISSUE TITLE
Compilation on MinGW is broken

### ISSUE THREAD

**Issue by gunterkoenigsmann** [wxMaxima-developers/wxmaxima#1315]
Seems like our unit tests currently don't compile on mingw

````
Q 98%] Building CXX object test/unit_tests/CMakeFiles/test_ImgCell.dir/test_ImgCell.cpp.obj
In file included from /usr/share/mingw-w64/include/windows.h:72:0,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/test/catch2/catch.hpp:7976,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/test/unit_tests/test_ImgCell.cpp:23:
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h: In function 'HWND__* CreateDialog(HINSTANCE, LPCTSTR, HWND, DLGPROC)':
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h:46:20: error: cannot convert 'LPCTSTR {aka const char*}' to 'LPCWSTR {aka const wchar_t*}' for argument '2' to 'HWND__* CreateDialogParamW(HINSTANCE, LPCWSTR, HWND, DLGPROC, LPARAM)'
             return CreateDialogW(hInstance, pTemplate, hwndParent, pDlgProc);
                    ^
In file included from /home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/defs.h:33:0,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/log.h:14,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/src/CellPtr.h:36,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/src/Cell.h:32,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/src/Cell.cpp:30,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/test/unit_tests/test_ImgCell.cpp:26:
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h: In function 'HFONT__* CreateFont(int, int, int, int, int, DWORD, DWORD, DWORD, DWORD, DWORD, DWORD, DWORD, DWORD, LPCTSTR)':
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h:77:48: error: cannot convert 'LPCTSTR {aka const char*}' to 'LPCWSTR {aka const wchar_t*}' for argument '14' to 'HFONT__* CreateFontW(int, int, int, int, int, DWORD, DWORD, DWORD, DWORD, DWORD, DWORD, DWORD, DWORD, LPCWSTR)'
                                family, facename);
                                                ^
In file included from /usr/share/mingw-w64/include/windows.h:72:0,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/test/catch2/catch.hpp:7976,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/test/unit_tests/test_ImgCell.cpp:23:
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h: In function 'HWND__* CreateWindow(LPCTSTR, LPCTSTR, DWORD, int, int, int, int, HWND, HMENU, HINSTANCE, LPVOID)':
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h:102:20: error: cannot convert 'LPCTSTR {aka const char*}' to 'LPCWSTR {aka const wchar_t*}' for argument '2' to 'HWND__* CreateWindowExW(DWORD, LPCWSTR, LPCWSTR, DWORD, int, int, int, int, HWND, HMENU, HINSTANCE, LPVOID)'
             return CreateWindowW(lpClassName, lpWndClass, dwStyle, x, y, w, h,
                    ^
In file included from /home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/defs.h:33:0,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/log.h:14,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/src/CellPtr.h:36,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/src/Cell.h:32,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/src/Cell.cpp:30,
                 from /home/maxima/maxima-code/crosscompile-windows/build/wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git/test/unit_tests/test_ImgCell.cpp:26:
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h: In function 'HMENU__* LoadMenu(HINSTANCE, LPCTSTR)':
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h:119:44: error: cannot convert 'LPCTSTR {aka const char*}' to 'LPCWSTR {aka const wchar_t*}' for argument '2' to 'HMENU__* LoadMenuW(HINSTANCE, LPCWSTR)'
             return LoadMenuW(instance, name);
                                            ^
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h: In function 'HICON__* LoadIcon(HINSTANCE, LPCTSTR)':
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h:310:51: error: cannot convert 'LPCTSTR {aka const char*}' to 'LPCWSTR {aka const wchar_t*}' for argument '2' to 'HICON__* LoadIconW(HINSTANCE, LPCWSTR)'
             return LoadIconW(hInstance, lpIconName);
                                                   ^
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h: In function 'HBITMAP__* LoadBitmap(HINSTANCE, LPCTSTR)':
/home/maxima/maxima-code/crosscompile-windows/build/wxwidgets/wxwidgets-prefix/src/wxwidgets/include/wx/msw/winundef.h:323:55: error: cannot convert 'LPCTSTR {aka const char*}' to 'LPCWSTR {aka const wchar_t*}' for argument '2' to 'HBITMAP__* LoadBitmapW(HINSTANCE, LPCWSTR)'
             return LoadBitmapW(hInstance, lpBitmapName);
                                                       ^
test/unit_tests/CMakeFiles/test_ImgCell.dir/build.make:63: recipe for target 'test/unit_tests/CMakeFiles/test_ImgCell.dir/test_ImgCell.cpp.obj' failed
make[5]: *** [test/unit_tests/CMakeFiles/test_ImgCell.dir/test_ImgCell.cpp.obj] Error 1
CMakeFiles/Makefile2:7042: recipe for target 'test/unit_tests/CMakeFiles/test_ImgCell.dir/all' failed
make[4]: *** [test/unit_tests/CMakeFiles/test_ImgCell.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make[3]: *** [all] Error 2
wxmaxima/CMakeFiles/wxmaxima-git.dir/build.make:111: recipe for target 'wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git-stamp/wxmaxima-git-build' failed
make[2]: *** [wxmaxima/wxmaxima-git-prefix/src/wxmaxima-git-stamp/wxmaxima-git-build] Error 2
CMakeFiles/Makefile2:556: recipe for target 'wxmaxima/CMakeFiles/wxmaxima-git.dir/all' failed
make[1]: *** [wxmaxima/CMakeFiles/wxmaxima-git.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2


Build information
-----------------

Operating system: 
Description:	Ubuntu 18.04.4 LTS
Maxima GIT Version: branch-5_44-base-18-g3a89c0dad
Wxmaxima GIT Version: Version-20.06.6-105-gc3aca9724
Build date and time (UTC): Tue Jun 23 22:31:13 UTC 2020
````

@KubaO :  I am still waiting for my laptop accumulator to be shipped. If you have time to look into this issue today I therefore would assign it to you. In theory the accumulator should have arrived until yesterday but in the currentbwild times currently the post is a bit unpredictable: Sometimes international parcels are extremely fast. Sometimes national ones last a week.

**Comment by KubaO** [comment]
I have 0 time today. But can look later this week. It is my problem of course so feel free to assign to me if you want.

**Comment by gunterkoenigsmann** [comment]
If we both are currently unavailable this is a ordinary task that will be done by the one who both has the necessary time and an idea on how to solve it at hands.

**Comment by gunterkoenigsmann** [comment]
Well... ...as a 2nd rhought: since thos is about string handling resolving this problem the Right Way might require your knowledge, though.

**Comment by KubaO** [comment]
It looks like include order problem. This test is the only one with that problem, and I imagine that the Catch runner includes things before wxwidgets had a change to insert itself in the mess known as Windows includes.

We need to add a CI build that does this cross compilation via mingw. Otherwise this will keep getting broken, I'm afraid.

Hopefully https://github.com/wxMaxima-developers/wxmaxima/tree/feature/fix-mingw-build will fix it. If so, please let me know so I'll merge it.

**Comment by gunterkoenigsmann** [comment]
Until the Appveyor MinGW build starts to work the only hint we have if MinGW works is Wolfgang's Nightly Build on https://wolfgang.dautermann.at. Since we cannot break this build right now I personnally would dare to merge fix-mingw-build hoping that this helps. But as we sometimes disagree in such things I'll let you decide.

**Comment by KubaO** [comment]
Is Wolfgang's nightly build something that can be run elsewhere? Is that the Docker image build? If so, either one of us can easily spin it up and see what it does.

I've not merged the fix branch since I don't know if it actually fixes it, but it's just a header reordering. It doesn't alter anything else. We may have nothing to lose by merging, but I'd rather run the build myself in Docker if that is available.

**Comment by daute** [comment]
The code for the nightly crosscompiled build is part of the Maxima project:
https://sourceforge.net/p/maxima/code/ci/master/tree/crosscompile-windows/nightlybuild.sh

Best regards, Wolfgang

**Comment by KubaO** [comment]
Thank you, Wolfgang. But this looks like just building Maxima. Where do the wxMaxima sources come from, and where it's being built? Does Maxima pull wxMaxima as a git submodule?

In any case: this is extremely vague, because to reproduce I need an exact build environment. For example: it's likely that this problem can be caused by an older version of MinGW - but which older version?? So, if Maxima builds wxMaxima on someone's personal computer or something like that, I have no hope of fixing anything reliably. It'd either need to be a Docker image, or an actual Travis or AppVeyor setup. There's a docker something-or-another on wxMaxima-developers org - is that somehow related?

**Comment by gunterkoenigsmann** [comment]
The docker images create appimage builds of maxima+wxMaxima for Linux. Wolfgang's Build for Windows is done on a real computer running Ubuntu instead.

**Comment by gunterkoenigsmann** [comment]
@daute :  The new error message looks like a necessary library beiing not installed on your computer:

````
usr/bin/x86_64-w64-mingw32-ld: cannot find -lWs2_32
collect2: error: ld returned 1 exit status
````
and:
````
[ 54%] Linking CXX executable wxmaxima.exe
/usr/bin/i686-w64-mingw32-ld: cannot find -lWs2_32
````
An attempt of googling this issue seems to indicate that minGW should come with this library and that it is the winsock library...

**Comment by daute** [comment]
The library names are case sensitive when crosscompiling on Linux and it is called "ws2_32" (lower case "W").

Fixed with commit  3833dd0.

(@KubaO: I am building a complete installer with Maxima, gnuplot, wxMaxima, etc. included). In
https://sourceforge.net/p/maxima/code/ci/master/tree/crosscompile-windows/wxmaxima/CMakeLists.txt
I do a git checkout of the wxMaxima sources and crosscompile the wxMaxima.)
